// corde/lib/src/types/general.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.permissionsArray =
  exports.Permission =
  exports.Colors =
  exports.ColorsHex =
  exports.SystemChannelFlag =
  exports.GuildFeatures =
  exports.VerificationLevel =
    void 0;
var VerificationLevel;

(function (VerificationLevel) {
  VerificationLevel["NONE"] = "NONE";
  VerificationLevel["LOW"] = "LOW";
  VerificationLevel["MEDIUM"] = "MEDIUM";
  VerificationLevel["HIGH"] = "HIGH";
  VerificationLevel["VERY_HIGH"] = "VERY_HIGH";
})((VerificationLevel = exports.VerificationLevel || (exports.VerificationLevel = {})));

var GuildFeatures;

(function (GuildFeatures) {
  GuildFeatures["INVITE_SPLASH"] = "INVITE_SPLASH";
  GuildFeatures["VIP_REGIONS"] = "VIP_REGIONS";
  GuildFeatures["VANITY_URL"] = "VANITY_URL";
  GuildFeatures["VERIFIED"] = "VERIFIED";
  GuildFeatures["PARTNERED"] = "PARTNERED";
  GuildFeatures["COMMUNITY"] = "COMMUNITY";
  GuildFeatures["COMMERCE"] = "COMMERCE";
  GuildFeatures["NEWS"] = "NEWS";
  GuildFeatures["DISCOVERABLE"] = "DISCOVERABLE";
  GuildFeatures["FEATURABLE"] = "FEATURABLE";
  GuildFeatures["ANIMATED_ICON"] = "ANIMATED_ICON";
  GuildFeatures["BANNER"] = "BANNER";
  GuildFeatures["WELCOME_SCREEN_ENABLED"] = "WELCOME_SCREEN_ENABLED";
  GuildFeatures["MEMBER_VERIFICATION_GATE_ENABLED"] = "MEMBER_VERIFICATION_GATE_ENABLED";
  GuildFeatures["PREVIEW_ENABLED"] = "PREVIEW_ENABLED";
})((GuildFeatures = exports.GuildFeatures || (exports.GuildFeatures = {})));

var SystemChannelFlag;

(function (SystemChannelFlag) {
  SystemChannelFlag["SUPPRESS_JOIN_NOTIFICATIONS"] = "SUPPRESS_JOIN_NOTIFICATIONS";
  SystemChannelFlag["SUPPRESS_PREMIUM_SUBSCRIPTIONS"] = "SUPPRESS_PREMIUM_SUBSCRIPTIONS";
})((SystemChannelFlag = exports.SystemChannelFlag || (exports.SystemChannelFlag = {})));

var ColorsHex;

(function (ColorsHex) {
  ColorsHex[(ColorsHex["DEFAULT"] = 0)] = "DEFAULT";
  ColorsHex[(ColorsHex["WHITE"] = 16777215)] = "WHITE";
  ColorsHex[(ColorsHex["AQUA"] = 1752220)] = "AQUA";
  ColorsHex[(ColorsHex["GREEN"] = 3066993)] = "GREEN";
  ColorsHex[(ColorsHex["BLUE"] = 3447003)] = "BLUE";
  ColorsHex[(ColorsHex["YELLOW"] = 16776960)] = "YELLOW";
  ColorsHex[(ColorsHex["PURPLE"] = 10181046)] = "PURPLE";
  ColorsHex[(ColorsHex["LUMINOUS_VIVID_PINK"] = 15277667)] = "LUMINOUS_VIVID_PINK";
  ColorsHex[(ColorsHex["GOLD"] = 15844367)] = "GOLD";
  ColorsHex[(ColorsHex["ORANGE"] = 15105570)] = "ORANGE";
  ColorsHex[(ColorsHex["RED"] = 15158332)] = "RED";
  ColorsHex[(ColorsHex["GREY"] = 9807270)] = "GREY";
  ColorsHex[(ColorsHex["NAVY"] = 3426654)] = "NAVY";
  ColorsHex[(ColorsHex["DARK_AQUA"] = 1146986)] = "DARK_AQUA";
  ColorsHex[(ColorsHex["DARK_GREEN"] = 2067276)] = "DARK_GREEN";
  ColorsHex[(ColorsHex["DARK_BLUE"] = 2123412)] = "DARK_BLUE";
  ColorsHex[(ColorsHex["DARK_PURPLE"] = 7419530)] = "DARK_PURPLE";
  ColorsHex[(ColorsHex["DARK_VIVID_PINK"] = 11342935)] = "DARK_VIVID_PINK";
  ColorsHex[(ColorsHex["DARK_GOLD"] = 12745742)] = "DARK_GOLD";
  ColorsHex[(ColorsHex["DARK_ORANGE"] = 11027200)] = "DARK_ORANGE";
  ColorsHex[(ColorsHex["DARK_RED"] = 10038562)] = "DARK_RED";
  ColorsHex[(ColorsHex["DARK_GREY"] = 9936031)] = "DARK_GREY";
  ColorsHex[(ColorsHex["DARKER_GREY"] = 8359053)] = "DARKER_GREY";
  ColorsHex[(ColorsHex["LIGHT_GREY"] = 12370112)] = "LIGHT_GREY";
  ColorsHex[(ColorsHex["DARK_NAVY"] = 2899536)] = "DARK_NAVY";
  ColorsHex[(ColorsHex["BLURPLE"] = 7506394)] = "BLURPLE";
  ColorsHex[(ColorsHex["GREYPLE"] = 10070709)] = "GREYPLE";
  ColorsHex[(ColorsHex["DARK_BUT_NOT_BLACK"] = 2895667)] = "DARK_BUT_NOT_BLACK";
  ColorsHex[(ColorsHex["NOT_QUITE_BLACK"] = 2303786)] = "NOT_QUITE_BLACK";
})((ColorsHex = exports.ColorsHex || (exports.ColorsHex = {})));

var Colors;

(function (Colors) {
  Colors["DEFAULT"] = "#000000";
  Colors["WHITE"] = "#FFFFFF";
  Colors["AQUA"] = "#1ABC9C";
  Colors["GREEN"] = "#2ECC71";
  Colors["BLUE"] = "#3498DB";
  Colors["PURPLE"] = "#9B59B6";
  Colors["GOLD"] = "#E91E63";
  Colors["ORANGE"] = "#E67E22";
  Colors["RED"] = "#E74C3C";
  Colors["GREY"] = "#95A5A6";
  Colors["DARKER_GREY"] = "#7F8C8D";
  Colors["NAVY"] = "#34495E";
  Colors["DARK_AQUA"] = "#11806A";
  Colors["DARK_GREEN"] = "#1F8B4C";
  Colors["DARK_BLUE"] = "#206694";
  Colors["DARK_PURPLE"] = "#71368A";
  Colors["DARK_GOLD"] = "#C27C0E";
  Colors["DARK_ORANGE"] = "#A84300";
  Colors["DARK_RED"] = "#992D22";
  Colors["DARK_GREY"] = "#7F8C8D";
  Colors["LIGHT_GREY"] = "#BCC0C0";
  Colors["DARK_NAVY"] = "#2C3E50";
  Colors["LUMINOUS_VIVID_PINK"] = "#E91E63";
  Colors["DARK_VIVID_PINK"] = "#AD1457";
})((Colors = exports.Colors || (exports.Colors = {})));

var Permission;

(function (Permission) {
  Permission[(Permission["CREATE_INSTANT_INVITE"] = 1)] = "CREATE_INSTANT_INVITE";
  Permission[(Permission["KICK_MEMBERS"] = 2)] = "KICK_MEMBERS";
  Permission[(Permission["BAN_MEMBERS"] = 4)] = "BAN_MEMBERS";
  Permission[(Permission["ADMINISTRATOR"] = 8)] = "ADMINISTRATOR";
  Permission[(Permission["MANAGE_CHANNELS"] = 16)] = "MANAGE_CHANNELS";
  Permission[(Permission["MANAGE_GUILD"] = 32)] = "MANAGE_GUILD";
  Permission[(Permission["ADD_REACTIONS"] = 64)] = "ADD_REACTIONS";
  Permission[(Permission["VIEW_AUDIT_LOG"] = 128)] = "VIEW_AUDIT_LOG";
  Permission[(Permission["VIEW_CHANNEL"] = 1024)] = "VIEW_CHANNEL";
  Permission[(Permission["SEND_MESSAGES"] = 2048)] = "SEND_MESSAGES";
  Permission[(Permission["SEND_TTS_MESSAGES"] = 4096)] = "SEND_TTS_MESSAGES";
  Permission[(Permission["MANAGE_MESSAGES"] = 8192)] = "MANAGE_MESSAGES";
  Permission[(Permission["EMBED_LINKS"] = 16384)] = "EMBED_LINKS";
  Permission[(Permission["ATTACH_FILES"] = 32768)] = "ATTACH_FILES";
  Permission[(Permission["READ_MESSAGE_HISTORY"] = 65536)] = "READ_MESSAGE_HISTORY";
  Permission[(Permission["MENTION_EVERYONE"] = 131072)] = "MENTION_EVERYONE";
  Permission[(Permission["USE_EXTERNAL_EMOJIS"] = 262144)] = "USE_EXTERNAL_EMOJIS";
  Permission[(Permission["CONNECT"] = 1048576)] = "CONNECT";
  Permission[(Permission["SPEAK"] = 2097152)] = "SPEAK";
  Permission[(Permission["MUTE_MEMBERS"] = 4194304)] = "MUTE_MEMBERS";
  Permission[(Permission["DEAFEN_MEMBERS"] = 8388608)] = "DEAFEN_MEMBERS";
  Permission[(Permission["MOVE_MEMBERS"] = 16777216)] = "MOVE_MEMBERS";
  Permission[(Permission["USE_VAD"] = 33554432)] = "USE_VAD";
  Permission[(Permission["PRIORITY_SPEAKER"] = 256)] = "PRIORITY_SPEAKER";
  Permission[(Permission["CHANGE_NICKNAME"] = 67108864)] = "CHANGE_NICKNAME";
  Permission[(Permission["MANAGE_NICKNAMES"] = 134217728)] = "MANAGE_NICKNAMES";
  Permission[(Permission["MANAGE_ROLES"] = 268435456)] = "MANAGE_ROLES";
  Permission[(Permission["MANAGE_WEBHOOKS"] = 536870912)] = "MANAGE_WEBHOOKS";
  Permission[(Permission["MANAGE_EMOJIS"] = 1073741824)] = "MANAGE_EMOJIS";
  Permission[(Permission["STREAM"] = 512)] = "STREAM";
  Permission[(Permission["VIEW_GUILD_INSIGHTS"] = 524288)] = "VIEW_GUILD_INSIGHTS";
})((Permission = exports.Permission || (exports.Permission = {})));

const permissionsArray = Object.keys(Permission);
exports.permissionsArray = permissionsArray;
