// corde/lib/src/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.bot =
  exports.beforeEach =
  exports.afterEach =
  exports.beforeStart =
  exports.afterAll =
  exports.expect =
  exports.it =
  exports.describe =
  exports.test =
  exports.group =
    void 0;

const tslib_1 = require("tslib");

var closures_1 = require("./closures");

Object.defineProperty(exports, "group", {
  enumerable: true,
  get: function () {
    return closures_1.group;
  },
});
Object.defineProperty(exports, "test", {
  enumerable: true,
  get: function () {
    return closures_1.test;
  },
});
Object.defineProperty(exports, "describe", {
  enumerable: true,
  get: function () {
    return closures_1.group;
  },
});
Object.defineProperty(exports, "it", {
  enumerable: true,
  get: function () {
    return closures_1.test;
  },
});

var expect_1 = require("./expect");

Object.defineProperty(exports, "expect", {
  enumerable: true,
  get: function () {
    return expect_1.expect;
  },
});

var hooks_1 = require("./hooks");

Object.defineProperty(exports, "afterAll", {
  enumerable: true,
  get: function () {
    return hooks_1.afterAll;
  },
});
Object.defineProperty(exports, "beforeStart", {
  enumerable: true,
  get: function () {
    return hooks_1.beforeStart;
  },
});
Object.defineProperty(exports, "afterEach", {
  enumerable: true,
  get: function () {
    return hooks_1.afterEach;
  },
});
Object.defineProperty(exports, "beforeEach", {
  enumerable: true,
  get: function () {
    return hooks_1.beforeEach;
  },
});
(0, tslib_1.__exportStar)(require("./types/general"), exports);

const corde_1 = require("./corde");

exports.bot = corde_1.corde.bot;
exports.default = corde_1.corde;
